### PR TITLE
Implement frequentist analysis for revenue test

### PR DESCRIPTION
**Pull request by valentino-amadeus**
#### What is the purpose of this pull request?
- Implement math functions for the frequentist analysis of the `revenue` type test.
- Call the appropriate functions, when performing the analysis, accordingly to the type of test.

#### What problem is this solving?
From now on, the tests of type `revenue` are going to be run as they actually should - as tests of type `revenue`.

#### How should this be manually tested?
Initialize a test, get some traffic and call the `status` and the `finish` routes.

#### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Requires change to documentation, which has been updated accordingly.

### CODE REVIEW

File: node/abTest/analysis/compareWorkspaces/index.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { EvaluateConversion } from './conversion'
+import { EvaluateRevenue } from './revenue'
+
+export async function Evaluate(testType: TestType, abTestBeginning: string, workspaceAData: WorkspaceCompleteData, workspaceBData: WorkspaceCompleteData): Promise<TestResult> {
+
+    if (testType === 'revenue') {
+        return await EvaluateRevenue(abTestBeginning, workspaceAData, workspaceBData)
+    }
+    return await EvaluateConversion(abTestBeginning, workspaceAData, workspaceBData)
```
Comment (tergol1) [inline node/abTest/analysis/compareWorkspaces/index.ts:9]: Same thing here about using a switch for test type

**Review by tergol1 — APPROVED** [review]
LGTM

File: node/abTest/analysis/compareWorkspaces/revenue.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    // For conversion evaluation
+    const betaDistributionA = WorkspaceToBetaDistribution(workspaceAData.SinceBeginning)
+    const betaDistributionB = WorkspaceToBetaDistribution(workspaceBData.SinceBeginning)
+    const lossA = LossFunctionChoosingVariantOne(betaDistributionA, betaDistributionB)
+    const lossB = LossFunctionChoosingVariantOne(betaDistributionB, betaDistributionA)
+    const probabilityTwoBeatsOne = ProbabilityOfOneBeatsTwo(betaDistributionB.a, betaDistributionB.b, betaDistributionA.a, betaDistributionA.b)
+    const PValueConversion = pValueConversion(betaDistributionA, betaDistributionB)
+    const winner = ChooseWinner(workspaceAData.SinceBeginning, workspaceBData.SinceBeginning, BoundError) || 'Not yet decided'
```
[the remark this answers is not in the document]
Comment (valentino-amadeus) [reply node/abTest/analysis/compareWorkspaces/revenue.ts:26]: In the latest branch, there are several places where this kind of change is more than welcome. Despite all the refinements in code structuring, the recent addition of functionalities was so big that the code still needs improvements. In my opinion, `ab-tester` needs a complete refactoring. But it doesn't make sense to do that now. We shall first resolve the analysis - and the splitting into calculations and traffic division.

File: node/abTest/analysis/compareWorkspaces/revenue.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    // For revenue evaluation
+    const uValue = CalculateUValue([ordersHistoryA, ordersHistoryB], 0)
+    const pValueRevenue = PValueRevenue(uValue, ordersHistoryA.length, ordersHistoryB.length)
+    const winnerRevenue = PickWinner(pValueRevenue, workspaceAData.SinceBeginning.Workspace, workspaceBData.SinceBeginning.Workspace)
+    const effectSizeA = CalculateEffectSize(uValue, ordersHistoryA.length, ordersHistoryB.length)
+    const effectSizeB = 1 - effectSizeA
+    const medianA = CalculateMedian(workspaceAData.SinceBeginning.OrdersValueHistory)
+    const medianB = CalculateMedian(workspaceBData.SinceBeginning.OrdersValueHistory)
```
[the remark this answers is not in the document]
Comment (valentino-amadeus) [reply node/abTest/analysis/compareWorkspaces/revenue.ts:35]: >we can almost always substitute comments in the code by functions that says what we're doing by its name.
Yeah, that's something I have learned in the last months. In that specific case, I don't think the result was bad, though. 🤔

File: node/utils/evaluation-response.ts [hunk 4]
```diff
@@ -1,3 +1,12 @@
+export const DefaultEvaluationResponse = (abTestBeginning: string, workspaceAName: string, workspaceBName: string, testType: TestType): TestResult => {
+    
+    if (testType === 'revenue') {
+        return DefaultEvaluationResponseRevenue(abTestBeginning, workspaceAName, workspaceBName)
+    }
+
+    return DefaultEvaluationResponseConversion(abTestBeginning, workspaceAName, workspaceBName)
```
[the remark this answers is not in the document]
Comment (valentino-amadeus) [reply node/utils/evaluation-response.ts:7]: If we happen to have more cases, a switch will be welcome. There is no reason for doing it now, as we only have one `if` - the default result has to be `DefaultEvaluationResponseConversion` for now, as there still are tests for which `testType = undefined`.

**Comment by valentino-amadeus** [comment]
Have you checked whether these changes introduce any unwanted behavior? These PRs are considerably old and some of them may have small problems. Although I tried to be very explicit about the presence of breaking changes (if I haven't specified that the PR contains any, it likely doesn't), it is always good to check if any of the later PRs fixes a bug related with the one you are reviewing. The latest state of the app (the one from the latest branch) is the one I rely the most on, but there might be intermediary branches that depend on others.

File: node/abTest/analysis/compareWorkspaces/revenue.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (tergol1) [reply node/abTest/analysis/compareWorkspaces/revenue.ts:37]: Not a bad result at all! I think that the only possible improvement would be exchanging comment + code for a helper which contains a name that specifies what is being done, what would make the comment unnecessary.

**Pull request merged**